src/transactions.rs

```diff
@@ -0,0 +1,456 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Seahorse library.
+
+// This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
+// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+// You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! The assets module.
+//!
+//! This module defines [Transaction], [TransactionEditor], and [Transactions], which provide CURD (create, read,
+//! update, and delete) operations, with the use of [KeyValueStore] to control the assets resource.
+
+use crate::{
+    key_value_store::*, AssetCode, KeystoreError, Ledger, RecordAmount, TransactionHash,
+    TransactionKind, TransactionReceipt, TransactionUID,
+};
+use chrono::{DateTime, Local};
+use jf_cap::{
+    keys::UserAddress,
+    structs::{ReceiverMemo, RecordOpening},
+    Signature,
+};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use std::collections::{BTreeMap, HashMap, HashSet};
+use std::ops::{Deref, DerefMut};
+
+/// A Transaction<L>with its UID as the primary key.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct Transaction<L: Ledger> {
+    /// Identifier for the Transaction, also the primary key in storage
+    uid: TransactionUID<L>,
+    /// Time when this transaction will expire if it's not completed
+    timeout: u64,
+    /// Hash which appears in the commited block for this transaction
+    hash: Option<TransactionHash<L>>,
```

Comment:
The reason this is `Option` is actually a lot dumber than that. It's supposed to be a hash of a `Transaction<L>` but when we create the `TransactionHistoryEntry` in the transaction builder we only have a CAP `TransactionNote`. We can't do `Transaction::<L>::cap(note.clone()).hash()` either, because e.g. in CAPE we call `transfer` to generate a burn note and then turn it into a `Transaction<CapeLedger>` using `CapeTransition::Burn`, not `CapeTransition::cap`.

So, I think it needs to stay this way unfortunately, just note in the comments that the hash gets set on submission, we don't have to wait for it to be committed

Reply:
Actually, maybe a reasonable approach would be to set the hash to `Transaction::<L>::cap(note.clone()).hash()` after building it, but make the hash public so the caller can update it if they want before calling `Keystore::submit`. That way you get a good default, the hash doesn't have to be an `Option` which probably saves us some unwraps or error handling that will never happen, and we can still do tricks like `burn`.

Reply:
I took more of a look into this and realized that `Transaction::<L>::cap` takes 2 parameters: the Note and the nullifier proofs.  To get the proofs we have to request them from the backend.  It looks like we do this right before we submit the transaction.  

I kept the hash optional because I think in both cases, Cap and Custom (like CAPE burn), we don't have all the information we need to generate the hash until right before we submit.  I am open to reconsider this but I think it would require a bit more refactoring to have the nullifier proofs available in `txn_builder` when we first store the transaction and I'm not sure that's desirable. 

I think for now I'll update the comments to reflect that we don't need to get a block to get the transaction hash we should update the hash before we call `submit`

Reply:
Ah, that makes sense. I still think we should refactor this eventually but agree it is too much work for right now. Can you make an issue so we can come back to it in the future?

Reply:
https://github.com/EspressoSystems/seahorse/issues/217